[MOD] <stock_cancel> reverting quant

oihane opened the pull request.

oihane committed:
[MOD] <stock_cancel> reverting quant

oihane committed:
[IMP] Added test to <stock_cancel>

pedrobaeza commented on stock_cancel/models/stock.py:
```diff
@@ -47,15 +48,27 @@ def action_revert_done(self):
             if picking.invoice_id:
                 raise exceptions.Warning(
                     _('Picking %s has invoices!') % (picking.name))
+            for move in picking.move_lines:
```
You can save one loop with: `for quant in picking.mapped('move_lines.quant_ids'):`

oihane replied:
No, I can't take a look to line 64

pedrobaeza commented on stock_cancel/models/stock.py:
```diff
@@ -47,15 +48,27 @@ def action_revert_done(self):
             if picking.invoice_id:
                 raise exceptions.Warning(
                     _('Picking %s has invoices!') % (picking.name))
```
Better "Picking has been invoiced"

pedrobaeza commented on stock_cancel/models/stock.py:
```diff
@@ -47,15 +48,27 @@ def action_revert_done(self):
+                quant_obj.quants_unreserve(move)
             picking.move_lines.write({'state': 'draft'})
             picking.state = 'draft'
             if picking.invoice_state == 'invoiced' and not picking.invoice_id:
```
You don't need to check again `not picking.invoice_id`, as this is a warning condition

pedrobaeza commented on stock_cancel/tests/test_stock_cancel.py:
```diff
@@ -0,0 +1,62 @@
+    def test_transferred_stock_picking_cancel(self):
+        self.assertEqual(self.picking_in.state, 'done')
+        self.picking_in.action_revert_done()
+        self.assertEqual(self.picking_in.state, 'draft')
```
Test here also the stock level and quant locations

oihane committed:
[IMP] <stock_cancel> Improving functionality
to fix errors when cancelling pickings with more than one move and
one of them can not be reverted

oihane committed:
[IMP] <stock_cancel> Updates in README and version

oihane committed:
[IMP] <stock_cancel> Delete quant when reverted picking is incoming type

wdbaruni commented:
Any updates with this fix?

pedrobaeza commented on stock_cancel/models/stock.py:
```diff
@@ -46,16 +47,32 @@ def action_revert_done(self):
-                    _('Picking %s has invoices!') % (picking.name))
+                    _('Picking %s has been invoiced') % (picking.name))
+            for move in picking.move_lines:
+                quants = move.quant_ids.filtered(
```
Why do you need to make this filter? With it, you're discarding associated quants that can be considered as a stopper for making the cancel, don't you think?

clonedagain replied:
Also it won't work in multi-location databases, where the actual location of the quant is decided using pack operations not stock moves (using manual entry and automatic put-away strategy).
Usually you have:
- move dest = the warehouse's location
- pack operation dest = the actual place where the goods were put.

clonedagain commented on stock_cancel/models/stock.py:
```diff
@@ -46,16 +47,32 @@ def action_revert_done(self):
+            for move in picking.move_lines:
+                quants = move.quant_ids.filtered(
+                    lambda q: q.location_id == move.location_dest_id)
+                if quants.filtered(lambda q: q.qty != move.product_uom_qty):
```
What if a single move touched several quants? You'll refuse to undo it won't you?

clonedagain commented on stock_cancel/tests/test_stock_cancel.py:
```diff
@@ -0,0 +1,74 @@
+                'product_uom_qty': 288,
+                'product_uom': self.product_1.uom_id.id,
+                'location_id': self.supplier_location.id,
+                'location_dest_id': self.stock_location.id,
```
You should add several lines with the same product and different lots

clonedagain commented on stock_cancel/tests/test_stock_cancel.py:
```diff
@@ -0,0 +1,74 @@
+            })],
+        })
+        self.picking_in.action_confirm()
+        self.picking_in.do_prepare_partial()
```
Here you must simulate what happens when users manually put away the goods in another location than the one on the stock move.

clonedagain commented:
:-1: at the moment I'm not convinced what you're trying to achieve is feasible, or at least reasonable.
There is a way to amend a picking in the standard, so I think this is too risky for the little good it does.

Please add tests to prove me wrong, I'll be glad to be mistaken. Messing with quants is not something we should do without **very** extensive testing.

Cedric-Pigeon commented:
@oihane  Hello, do you plan to achieve this PR ? Is it still usefull ? Otherwise it will be closed in 3 days. Thanks.

oihane commented:
@Cedric-Pigeon you can close it, I have not done anything since 2015 with this PR.
